controllers/openstackdataplanedeployment_controller.go

```diff
@@ -0,0 +1,274 @@
+	for _, nodeSet := range nodeSets.Items {
+		if !nodeSet.Status.Conditions.IsTrue(dataplanev1.SetupReadyCondition) {
+			logger.Info("NodeSet SetupReadyCondition is not True", "NodeSet", nodeSet.Name)
+			return ctrl.Result{RequeueAfter: time.Second * 15}, nil
+		}
+	}
+
+	// All nodeSets successfully fetched.
+	// Mark InputReadyCondition=True
+	instance.Status.Conditions.MarkTrue(condition.InputReadyCondition, condition.ReadyMessage)
+
+	// Deploy each nodeSet
+	// The loop starts and checks NodeSet deployments sequentially. However, after they
+	// are started, they are running in parallel, since the loop does not wait
+	// for the first started NodeSet to finish before starting the next.
+	shouldRequeue := false
+	haveError := false
+	for _, nodeSet := range nodeSets.Items {
+
+		logger.Info(fmt.Sprintf("Deploying NodeSet: %s", nodeSet.Name))
+		logger.Info("Set Status.Deployed to false", "instance", instance)
+		instance.Status.Deployed = false
+		logger.Info("Set DeploymentReadyCondition false", "instance", instance)
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.DeploymentReadyCondition, condition.RequestedReason,
+			condition.SeverityInfo, condition.DeploymentReadyRunningMessage))
+		ansibleEESpec := nodeSet.GetAnsibleEESpec()
+		ansibleEESpec.AnsibleTags = instance.Spec.AnsibleTags
+		ansibleEESpec.AnsibleSkipTags = instance.Spec.AnsibleSkipTags
+		ansibleEESpec.AnsibleLimit = instance.Spec.AnsibleLimit
+
+		nodeSetSecretInv := fmt.Sprintf("dataplanenodeset-%s", nodeSet.Name)
+
+		if nodeSet.Status.DNSClusterAddresses != nil && nodeSet.Status.CtlplaneSearchDomain != "" {
+			ansibleEESpec.DNSConfig = &corev1.PodDNSConfig{
+				Nameservers: nodeSet.Status.DNSClusterAddresses,
+				Searches:    []string{nodeSet.Status.CtlplaneSearchDomain},
+			}
+		}
+
+		deployResult, err := deployment.Deploy(
+			ctx, helper, &nodeSet, instance,
+			nodeSetSecretInv, &instance.Status, ansibleEESpec,
+			nodeSet.Spec.Services)
+
+		if err != nil {
+			util.LogErrorForObject(helper, err, fmt.Sprintf("OpenStackDeployment error for NodeSet %s", nodeSet.Name), instance)
+			haveError = true
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				condition.ReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				dataplanev1.DataPlaneNodeSetErrorMessage,
+				err.Error()))
+		}
+
+		if deployResult != nil {
+			shouldRequeue = true
+		} else {
+			logger.Info("OpenStackDeployment succeeded for NodeSet", "NodeSet", nodeSet.Name)
+			logger.Info("Set NodeSetDeploymentReadyCondition true", "nodeSet", nodeSet.Name)
+			instance.Status.Conditions.Set(
+				condition.TrueCondition(
+					condition.Type(fmt.Sprintf(dataplanev1.NodeSetDeploymentReadyCondition, nodeSet.Name)),
+					condition.DeploymentReadyMessage))
+
+			logger.Info("Set NodeSet DeploymentReadyCondition true", "nodeSet", nodeSet.Name)
+			_, err = controllerutil.CreateOrPatch(ctx, helper.GetClient(), &nodeSet, func() error {
```

Comment:
I see, I can make this change. I suppose Watches aren't that resource intensive, or do we need to consider that cost at all? We would also have to filter the watch events for Deployments that actually have the NodeSet in their nodeSets list. That ought to be do-able in the watcher func.

Is this change blocking for you, or can I add it in a followup?

Reply:
> kubebuilder book seem to imply it's a fairly expensive operation on the cluster

I don't think it is, that too when there is one `openstackdataplanedeployment` resource per `openstackdataplanenodeset` at a time. We had this in the old implementation when node status was set when role is ready[1]. Nodeset still watches dnsmasq[2] and it's the standard pattern implemented across all operators[3].

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/dev-preview/controllers/openstackdataplanenode_controller.go#L311

[2] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/controllers/openstackdataplanenodeset_controller.go#L348-L349
[3] https://github.com/openstack-k8s-operators/infra-operator/blob/main/controllers/network/ipset_controller.go#L184

Reply:
@rabi Can you clarify about the anti-pattern? Where's the guidance on using Watch vs just using something simpler like CreateOrPatch? Perhaps CreateOrPatch is only intended for resources you might create? Although, there is also simpler Patch methods that would patch just the difference in the status subresource on the NodeSet.

I have added the Watch in a 2nd commit on this PR, but would like to understand better if its worth the added overhead.

Reply:
> @rabi Can you clarify about the anti-pattern? Where's the guidance on using Watch vs just using something simpler like CreateOrPatch? 

I think that is the core principle of controller design[1]. A controller is `only` responsible to `manage` resources it owns. Also, the `Status` of a CR is an outcome of a reconciliation and should not be patched externally with API or from another controller that does not own the CR. Attempting to update the status of a CR that a controller does not own can lead to issues and conflicts. Instead of trying to directly update the status of a CR that a controller does not own, implementing an event-driven approach is always recommended.


A bit more detail on controller responsibilities:

Controller Focus: Controller primarily focuses is managing the resources it owns or is responsible for. It watches and reconciles those resources to ensure they are in the desired state.

Event Handling: In addition to managing its own resources, a controller can also watch for events related to external resources. These events might be changes, updates, or creations of CRDs or other resources that are relevant to your controller's functionality.

Event Processing: When your controller detects an event related to an external resource, it can process that event and take appropriate actions. These actions might include updating the status of its own resources, creating new resources, or triggering other operations as needed.

> I have added the Watch in a 2nd commit on this PR, but would like to understand better if its worth the added overhead.

I'm not sure why we think it's an overhead. Everything is event driven when it comes to reconciliation. `Owns` and `For` are also `Watches` under the hood[2]. We can also filter events using predicates[3], if we want a controller to not respond to specific events.

[1] https://kubernetes.io/docs/concepts/architecture/controller/
[2] https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/builder/controller.go#L113
[3] https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate